File: backtest.mjs
import YahooFinance from 'yahoo-finance2';
import fs from 'fs';

const yahooFinance = new YahooFinance();

async function main() {
    // Read input from command line
    let input = { "GLD": 0.3, "SPY": 0.5, "IEF": 0.2 };
    if (process.argv[2]) {
        try {
            const parsed = JSON.parse(process.argv[2]);
            // Convert back to weights 0.0-1.0
            input = {};
            for (const [k, v] of Object.entries(parsed)) {
                input[k] = v;
            }
        } catch (e) {}
    }

    const tickers = Object.keys(input);
    const weights = Object.values(input);
    const totalWeight = weights.reduce((a, b) => a + b, 0);
    const normalizedWeights = weights.map(w => w / totalWeight);
    
    const startPeriod = '1970-01-01';
    const endPeriod = new Date().toISOString().split('T')[0];
    
    let allData = {};
    for (const ticker of tickers) {
        try {
            const result = await yahooFinance.historical(ticker, {
                period1: startPeriod,
                period2: endPeriod,
                interval: '1d'
            });
            allData[ticker] = result;
        } catch (e) {
            console.error(`Failed to fetch ${ticker}:`, e.message);
        }
    }
    
    // get dividend yield
    let divYieldPortfolio = 0;
    for (let i = 0; i < tickers.length; i++) {
        const ticker = tickers[i];
        const weight = normalizedWeights[i];
        try {
            const quote = await yahooFinance.quote(ticker);
            const dy = quote.trailingAnnualDividendYield || quote.dividendYield || 0;
            divYieldPortfolio += dy * weight;
        } catch (e) { }
    }
    
    // align dates
    let commonDates = [];
    if (tickers.length > 0 && allData[tickers[0]]) {
        let datesSet = new Set(allData[tickers[0]].map(d => d.date.toISOString().split('T')[0]));
        for (let i=1; i < tickers.length; i++) {
            const ticker = tickers[i];
            if (!allData[ticker]) continue;
            const newSet = new Set();
            for (const d of allData[ticker]) {
                const ds = d.date.toISOString().split('T')[0];
                if (datesSet.has(ds)) newSet.add(ds);
            }
            datesSet = newSet;
        }
        commonDates = Array.from(datesSet).sort();
    }
    
    if (commonDates.length === 0) {
        console.log(JSON.stringify({ error: "No common dates or no data found" }));
        return;
    }
    
    let portValues = [];
    let prevPrices = {};
    let filteredDates = [];
    
    for (const ds of commonDates) {
        let prices = {};
        let valid = true;
        for (const ticker of tickers) {
            const point = allData[ticker].find(d => d.date.toISOString().split('T')[0] === ds);
            if (!point || point.adjClose === undefined) { valid = false; break; }
            prices[ticker] = point.adjClose;
        }
        if (!valid) continue;
        
        filteredDates.push(ds);
        
        if (portValues.length === 0) {
            portValues.push(1.0); 
        } else {
            let dailyReturn = 0;
            for (let j=0; j < tickers.length; j++) {
                const ticker = tickers[j];
                const ret = (prices[ticker] - prevPrices[ticker]) / prevPrices[ticker];
                dailyReturn += ret * normalizedWeights[j];
            }
            portValues.push(portValues[portValues.length - 1] * (1 + dailyReturn));
        }
        prevPrices = prices;
    }
    
    const actualStart = filteredDates[0];
    const actualEnd = filteredDates[filteredDates.length - 1];
    const years = (new Date(actualEnd) - new Date(actualStart)) / (1000 * 60 * 60 * 24 * 365.25);
    
    const finalValue = portValues[portValues.length - 1];
    const cagr = Math.pow(finalValue, 1 / years) - 1;
    
    let dailyReturns = [];
    for (let i=1; i < portValues.length; i++) {
        dailyReturns.push((portValues[i] - portValues[i-1]) / portValues[i-1]);
    }
    
    const mean = dailyReturns.reduce((a,b)=>a+b, 0) / dailyReturns.length;
    const variance = dailyReturns.reduce((a,b)=>a + Math.pow(b - mean, 2), 0) / (dailyReturns.length - 1);
    const stdDev = Math.sqrt(variance);
    const volatility = stdDev * Math.sqrt(252);
    
    const riskFreeRate = 0.02;
    const sharpeRatio = (cagr - riskFreeRate) / volatility;
    
    let runningMax = portValues[0];
    let mdd = 0;
    for (let i=1; i < portValues.length; i++) {
        if (portValues[i] > runningMax) runningMax = portValues[i];
        const dd = (portValues[i] - runningMax) / runningMax;
        if (dd < mdd) mdd = dd;
    }

    // History data (12-20 points)
    const historyValues = portValues.map(v => v * 1000);
    const step = Math.max(1, Math.floor(historyValues.length / 12));
    let history = [];
    for (let i = 0; i < historyValues.length; i += step) {
        history.push({
            date: filteredDates[i],
            value: Number(historyValues[i].toFixed(2))
        });
    }
    if (history[history.length-1].date !== filteredDates[filteredDates.length-1]) {
        history.push({
            date: filteredDates[filteredDates.length-1],
            value: Number(historyValues[historyValues.length-1].toFixed(2))
        });
    }

    const radar = [
        { subject: "Attack", A: Math.min(100, Math.max(0, Math.round(cagr * 400))), fullMark: 100 },
        { subject: "Defense", A: Math.min(100, Math.max(0, Math.round((1 + mdd) * 100))), fullMark: 100 },
        { subject: "Volatility", A: Math.min(100, Math.max(0, Math.round(100 - volatility * 200))), fullMark: 100 },
        { subject: "Sharpe", A: Math.min(100, Math.max(0, Math.round(sharpeRatio * 40))), fullMark: 100 },
        { subject: "Dividend", A: Math.min(100, Math.max(0, Math.round(divYieldPortfolio * 1000))), fullMark: 100 },
    ];
    
    const result = {
        metrics: {
            cagr: Number((cagr * 100).toFixed(2)),
            mdd: Number((mdd * 100).toFixed(2)),
            sharpe: Number(sharpeRatio.toFixed(2)),
            period: `${actualStart} ~ ${actualEnd}`
        },
        radar: radar,
        history: history
    };
    
    console.log(JSON.stringify(result, null, 2));
}

main().catch(err => {
    console.log(JSON.stringify({ error: err.message }));
});

File: backtest.py
import json
import sys
import numpy as np
import pandas as pd
import yfinance as yf
from datetime import datetime

def run_backtest(tickers_weights):
    tickers = list(tickers_weights.keys())
    weights = np.array([tickers_weights[t] for t in tickers], dtype=float)
    weights = weights / np.sum(weights)

    # Date range
    start_date = "1970-01-01"
    end_date = datetime.now().strftime('%Y-%m-%d')

    try:
        # Fetch tickers + SPY for benchmark
        fetch_tickers = list(set(tickers + ["SPY"]))
        data = yf.download(fetch_tickers, start=start_date, end=end_date, progress=False, group_by='column', actions=True)
        
        if data.empty:
            return {"error": "No data found for tickers"}
        
        # Extract Adj Close and Dividends
        if len(fetch_tickers) > 1:
            adj_close = data['Adj Close'] if 'Adj Close' in data.columns.levels[0] else data['Close']
            dividends = data['Dividends'] if 'Dividends' in data.columns.levels[0] else pd.DataFrame(0, index=adj_close.index, columns=adj_close.columns)
        else:
            # Only happens if user input only SPY
            adj_close = (data['Adj Close'] if 'Adj Close' in data.columns else data['Close']).to_frame()
            adj_close.columns = fetch_tickers
            dividends = (data['Dividends'] if 'Dividends' in data.columns else pd.DataFrame(0, index=adj_close.index, columns=fetch_tickers)).to_frame()
            dividends.columns = fetch_tickers
            
    except Exception as e:
        return {"error": f"Data fetch failed: {str(e)}"}

    # Align Portfolio: Drop rows where any portfolio ticker is NaN
    # But keep SPY separate for a moment to ensure we have overlapping data
    port_adj_close = adj_close[tickers].dropna()
    if port_adj_close.empty:
        return {"error": "No overlapping data for portfolio tickers."}

    # Now align everything (Portfolio + SPY) to the portfolio's inception
    common_index = port_adj_close.index
    adj_close_clean = adj_close.loc[common_index].ffill().dropna() # ffill SPY if needed, though usually it exists
    dividends_clean = dividends.loc[adj_close_clean.index].fillna(0)

    actual_start = adj_close_clean.index.min()
    actual_end = adj_close_clean.index.max()
    
    days = (actual_end - actual_start).days
    years = days / 365.25
    if years < 0.01:
        return {"error": "Insufficient historical data."}

    # 1. Portfolio Performance
    port_daily_returns = adj_close_clean[tickers].pct_change().dropna()
    weights_series = pd.Series(weights, index=tickers)
    weights_aligned = weights_series.reindex(port_daily_returns.columns).values
    port_returns = port_daily_returns.dot(weights_aligned)
    port_cum_returns = (1 + port_returns).cumprod()

    # 2. Benchmark (SPY) Performance
    bench_daily_returns = adj_close_clean['SPY'].pct_change().dropna()
    bench_cum_returns = (1 + bench_daily_returns).cumprod()

    # Metrics Helper
    def get_metrics(cum_returns, daily_returns, div_series, price_series, weights_val=None):
        final_val = cum_returns.iloc[-1]
        cagr = (final_val ** (1 / years)) - 1
        vol = daily_returns.std() * np.sqrt(252)
        sharpe = (cagr - 0.02) / vol if vol > 0 else 0
        mdd_series = (cum_returns - cum_returns.cummax()) / cum_returns.cummax()
        mdd = mdd_series.min()
        mdd_date = mdd_series.idxmin().strftime('%Y')

        # Best Year
        annual_returns = (1 + daily_returns).resample('YE').prod() - 1
        best_year_val = annual_returns.max()
        best_year_date = annual_returns.idxmax().strftime('%Y')
        
        # Dividend Yield (last 1Y)
        one_year_ago = actual_end - pd.Timedelta(days=365)
        recent_divs = div_series[div_series.index >= one_year_ago]
        
        if weights_val is not None:
            # Portfolio Div Yield
            dy_port = 0
            for t in tickers:
                w = weights_val[t]
                p = price_series[t].iloc[-1]
                d = recent_divs[t].sum() if t in recent_divs.columns else 0
                dy_port += (d / p * w) if p > 0 else 0
            dy = dy_port
        else:
            # Single asset (SPY) Div Yield
            p = price_series.iloc[-1]
            d = recent_divs.sum()
            dy = (d / p) if p > 0 else 0

        return {
            "cagr": round(float(cagr * 100), 2),
            "mdd": round(float(mdd * 100), 2),
            "mdd_year": mdd_date,
            "sharpe": round(float(sharpe), 2),
            "volatility": round(float(vol * 100), 2),
            "dividend": round(float(dy * 100), 2),
            "best_year": {"year": best_year_date, "value": round(float(best_year_val * 100), 2)}
        }

    port_m = get_metrics(port_cum_returns, port_returns, dividends_clean[tickers], adj_close_clean[tickers], weights_series)
    bench_m = get_metrics(bench_cum_returns, bench_daily_returns, dividends_clean['SPY'], adj_close_clean['SPY'])

    # Radar data (Portfolio vs Benchmark)
    subjects = ["Attack", "Defense", "Volatility", "Sharpe", "Dividend"]
    radar = []
    
    # New Normalization factors based on user request
    def normalize(val, subject):
        if subject == "Attack": 
            # Max 30%, steps of 7.5% (7.5=25, 15=50, 22.5=75, 30=100)
            return min(100, max(0, (val / 30) * 100))
        if subject == "Dividend": 
            # Max 10%, steps of 2.5% (2.5=25, 5=50, 7.5=75, 10=100)
            return min(100, max(0, (val / 10) * 100))
        if subject == "Defense": 
            # Metric: 1500 / |MDD|. MDD 30% = 50 pts, MDD 15% = 100 pts.
            mdd_abs = abs(val)
            if mdd_abs == 0: return 100
            return min(100, max(0, 1500 / mdd_abs))
        if subject == "Sharpe": 
            # Max 1.5, steps for 0.5, 0.75, 1.0, 1.25, 1.5
            # We'll map 1.5 to 100.
            return min(100, max(0, (val / 1.5) * 100))
        if subject == "Volatility": 
            # Proposal: 100 - (Vol * 2.5). 0% Vol = 100, 20% Vol = 50, 40% Vol = 0.
            return min(100, max(0, 100 - (val * 2.5)))
        return 0

    for s in subjects:
        p_val = port_m['cagr'] if s == "Attack" else port_m['mdd'] if s == "Defense" else port_m['volatility'] if s == "Volatility" else port_m['sharpe'] if s == "Sharpe" else port_m['dividend']
        b_val = bench_m['cagr'] if s == "Attack" else bench_m['mdd'] if s == "Defense" else bench_m['volatility'] if s == "Volatility" else bench_m['sharpe'] if s == "Sharpe" else bench_m['dividend']
        
        radar.append({
            "subject": s,
            "A": int(normalize(p_val, s)), # Portfolio
            "B": int(normalize(b_val, s)), # Benchmark
            "fullMark": 100
        })

    # History chart (Portfolio only for simplicity, or we can add benchmark)
    history_values = (port_cum_returns * 1000)
    if len(history_values) > 20:
        indices = np.linspace(0, len(history_values) - 1, 20).astype(int)
        sampled_history = history_values.iloc[indices]
    else:
        sampled_history = history_values

    history_data = [{"date": actual_start.strftime('%Y-%m-%d'), "value": 1000.0}]
    for date, val in sampled_history.items():
        history_data.append({"date": date.strftime('%Y-%m-%d'), "value": round(float(val), 2)})

    return {
        "metrics": port_m,
        "benchmark_metrics": bench_m,
        "period": f"{actual_start.strftime('%Y.%m')} ~ {actual_end.strftime('%Y.%m')}",
        "radar": radar,
        "history": history_data
    }

if __name__ == "__main__":
    try:
        input_json = sys.argv[1] if len(sys.argv) > 1 else '{"SPY": 0.6, "TLT": 0.4}'
        print(json.dumps(run_backtest(json.loads(input_json))))
    except Exception as e:
        print(json.dumps({"error": str(e)}))
